src/threads/ThreadState.cpp

```diff
@@ -278,4 +311,17 @@ int Level::getGlobalThreadNum(faabric::Message* msg)
 {
     return msg->appindex();
 }
+
+std::string Level::toString()
+{
+    std::stringstream ss;
+    ss << "Level " << id << " depth=" << depth << " threads=" << numThreads
+       << " shared={ ";
+    for (int i = 0; i < nSharedVarOffsets; i++) {
+        ss << sharedVarOffsets[i] << " ";
+    }
+    ss << "}";
+
+    return ss.str();
+}
```

Comment:
This function is only used for trace logging purposes, so we don't need to worry about performance or efficiency.